src/app/shared/recent-utils.ts

```diff
@@ -0,0 +1,88 @@
+import { ErrorHandler } from '@angular/core';
+import { Context, Space } from 'ngx-fabric8-wit';
+import { Observable, ReplaySubject, Subject } from 'rxjs';
+import { ExtProfile, ProfileService } from '../profile/profile.service';
+
+export const RECENT_LENGTH: number = 8;
+
+export interface RecentData <T extends Context | Space> {
+  data: T[];
+  isSaveRequired: boolean;
+}
```

Comment:
Looks good overall, but I'd like to find a way around this `T extends Context | Space` thing - that's not a nice type constraint, because the end result is a very neutered form of "genericism". Ideally we want to make it so that this abstract class never needs to be touched or updated in any way simply to allow for a "Recents Store" of another data type than Spaces or Contexts. This type assertion technique also gets hairy if different datatypes end up with a lot of properties with the same name. Specific thoughts on potential solutions below on related bits of code.

Reply:
That sounds fair, I've taken your suggested route with adding a function `compareElements` to be implemented by the subclass for use in `findIndex`. The only problem I ran into is that the Context Service needs to respond to space deletion events, so in this case it isn't enough to have a comparator for type `T`, because the first element would be a space and the second would be a context. What I've done here is added an additional comparator function for the context service that will be passed into `onBroadcastSpaceDeleted`, which defaults to the implemented `compareElements` if a function is not passed to it.